File: 2024/day01_Historian_Hysteria/main.py
import sys
from collections import Counter


def part1(file: list[str]):
    l1, l2 = zip(*(line.split() for line in file))
    l1, l2 = sorted(l1), sorted(l2)

    res = sum([abs(int(l1[i]) - int(l2[i])) for i in range(len(l1))])
    
    print(res)

def part2(file: list[str]):
    l1, l2 = zip(*(line.split() for line in file))
    dict2 = Counter(l2)

    res = sum(int(num) * dict2[num] for num in l1 if num in dict2)

    print(res)


with open(sys.argv[1]+".txt", "r") as file:
    lines = [line.rstrip("\n") for line in file]

part1(lines)
part2(lines)

File: 2024/day01_Historian_Hysteria/src/main.rs
use std::io::stdin;


fn part1(file: &Vec<String>) {
    let pairs: Vec<(i32, i32)> = file
        .iter()
        .map(|line| {
            let mut iter = line.split("   ")
                .map(|x| {
                    x.parse::<i32>().unwrap()
                });
            (iter.next().unwrap(), iter.next().unwrap())
        })
        .collect();

    let (mut l1, mut l2): (Vec<i32>, Vec<i32>) = pairs.iter().cloned().unzip();
    l1.sort();
    l2.sort();


    let res = l1.iter().zip(l2.iter())
        .map(|(x, y)| (x - y).abs())
        .sum::<i32>();

    println!("{res}");
}

fn part2(file: &Vec<String>) {
    todo!();
}

fn main() {
    let mut file = Vec::new();
    loop {
        let mut line = String::new();
        match stdin().read_line(&mut line) {
            Ok(0) => break,
            Ok(_) => file.push(line.trim().to_string()),
            Err(_) => break,
        }
    }

    part1(&file);
    part2(&file);
}
